src/Revealer.tsx

```diff
@@ -0,0 +1,99 @@
+import React, { useState, useEffect, useRef } from "react";
+import { Typography, LinearProgress } from "@material-ui/core";
+import { RevealState, RoleCard } from "./RoleCard";
+
+export enum Team {
+  RESISTANCE = "Resistance",
+  SPY = "Spy"
+}
+
+export interface PlayerInfo {
+  name: string;
+  team: Team;
+}
+
+interface Props {
+  players: PlayerInfo[];
+}
+
+export const Revealer: React.FC<Props> = (props: Props) => {
+  const [currentPlayerIndex, setCurrentPlayerIndex] = useState<number>(0);
+  const [revealState, setRevealState] = useState<RevealState>(
+    RevealState.WAITING
+  );
+  const [revealProgress, setRevealProgress] = useState<number>(0);
+
+  const revealStart = (): void => {
+    if (revealState === RevealState.WAITING) {
+      setRevealState(RevealState.HOLDING);
+    }
+  };
+
+  const revealEnd = (): void => {
+    if (revealState === RevealState.HOLDING) {
+      setRevealState(RevealState.WAITING);
+      setRevealProgress(0);
+    } else if (revealState === RevealState.REVEALED) {
+      setRevealState(RevealState.PASSING);
+    }
+  };
+
+  const renderBody = (): React.ReactNode => {
+    const currentPlayer = props.players[currentPlayerIndex];
+    if (currentPlayer) {
+      return (
+        <RoleCard
+          playerName={currentPlayer.name}
+          team={currentPlayer.team}
+          revealState={revealState}
+          onRevealStart={revealStart}
+          onRevealEnd={revealEnd}
+        />
+      );
+    } else {
+      return <Typography>All set!</Typography>;
+    }
+  };
+
+  // This feels really ugly, but the refs seem to be necessary
+  // for anything that's read in the timer function
+  const currentPlayerIndexRef = useRef(currentPlayerIndex);
+  currentPlayerIndexRef.current = currentPlayerIndex;
+  const revealStateRef = useRef(revealState);
+  revealStateRef.current = revealState;
+  const revealProgressRef = useRef(revealProgress);
+  revealProgressRef.current = revealProgress;
+
+  const updateRevealProgress = (): void => {
+    if (revealStateRef.current === RevealState.HOLDING) {
+      if (revealProgressRef.current < 100) {
+        setRevealProgress(revealProgressRef.current + 50);
```

Comment:
I'd like to avoid using refs, but I don't think there's a way I can pass a function to update `revealProgress` here.

If `updateRevealProgress()` was simpler -- let's say it just always increased `revealProgress` by 50, then we could do:
```typescript
const updateRevealProgress = (): void => {
    setRevealProgress((revealProgress: number) => revealProgress + 50);
}
```

or if how we update `revealProgress` depended on the current value, that would be fine too:
```typescript
const updateRevealProgress = (): void => {
    setRevealProgress((rp: number) => rp === 100 ? rp - 25 : rp + 50);
}
```

But in this case how to update `revealProgress` also depends on `revealState`. Is there a different pattern here that might work?

Reply:
So here's some thoughts in no particular order.

You don't ever reset the `current`s inside this function, suggesting that there must be a way to define this function without relying on refs. Usually, refs are used because some side effect needs to write them, not simply read them.

The proper way to pass these state variables to these types of callbacks is by capturing them in the closure and stating them in the deps list (which I think you already know). There's two ways you can do that that aren't that different from what you do now:

- define this function inline in `useEffect`, and add the deps to `useEffect`
- define this function in `useCallback`, and make `useEffect` depend on the callback

In both cases, it's directly or indirectly making `useEffect` depend on the state variables. I think this is better, though it still feels like there's an even more correct pattern. I'm still pretty new to hooks, and trying to wrap my head around them a bit.

Doing the second option would look kind of like

```
const updateRevealProgress = useCallback(() => { ... }, [ deps ]);

// basic version
// this version might irregularly update, since it's always 500ms since the last change
useEffect(() => {
  const timer = window.setInterval(updateRevealProgress, 500);
  return () => { window.clearInterval(timer); };
}, [updateRevealProgress]);

// version that guarantees consistent ~500ms intervals
const lastTimerStart = useRef(Date.now());
useEffect(() => {
  const timeSinceLastTimer = Date.now() - lastTimerStart.current();
  lastTimerStart.current = Date.now();
  let interval: number | undefined;
  const timeout = window.setTimeout(() => {
    updateRevealProgress();
    interval = setInterval(updateRevealProgress, 500);
  }, Math.max(0, 500 - timeSinceLastTimer));
  return () => {
    window.clearTimeout(timeout);
    window.clearInterval(interval);
  };
}, [updateRevealProgress]);
```

This might be overengineering it, but there it is. Just some ideas that came to mind. I don't see a way to not use refs, which isn't surprising because timers are generally instance properties on class components, and refs are the functional version of instance properties.